File: xml-sdk/src/test/java/com/cleanster/xml/WebhooksTest.java
package com.cleanster.xml;

import com.cleanster.xml.api.WebhooksXmlApi;
import com.cleanster.xml.client.XmlConverter;
import com.cleanster.xml.client.XmlHttpClient;
import com.cleanster.xml.model.Webhook;
import com.cleanster.xml.model.XmlApiResponse;
import org.junit.jupiter.api.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

class WebhooksTest {

    private XmlHttpClient   http;
    private WebhooksXmlApi  api;

    private static final String LIST_JSON  = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":[{\"id\":1,\"url\":\"https://example.com/hook\",\"events\":[\"booking.completed\"]}]}";
    private static final String HOOK_JSON  = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":{\"id\":1,\"url\":\"https://example.com/hook\",\"events\":[\"booking.completed\"]}}";
    private static final String OK_JSON    = "{\"success\":true,\"message\":\"OK\",\"data\":{}}";

    @BeforeEach void setUp() {
        XmlHttpClient real = XmlHttpClient.builder().baseUrl("http://dummy").accessKey("key").build();
        http = spy(real);
        api  = new WebhooksXmlApi(http);
    }

    // ── listWebhooks ───────────────────────────────────────────────────────────

    @Test void listWebhooks_callsGet() {
        doReturn(LIST_JSON).when(http).get("/v1/webhooks");
        api.listWebhooks();
        verify(http).get("/v1/webhooks");
    }

    @Test void listWebhooks_returnsParsedList() {
        doReturn(LIST_JSON).when(http).get("/v1/webhooks");
        XmlApiResponse<List<Webhook>> resp = api.listWebhooks();
        assertTrue(resp.isSuccess());
        assertEquals(1, resp.getData().size());
    }

    @Test void listWebhooks_emptyList() {
        doReturn("{\"success\":true,\"message\":\"OK\",\"data\":[]}")
                .when(http).get("/v1/webhooks");
        assertTrue(api.listWebhooks().getData().isEmpty());
    }

    // ── createWebhook ──────────────────────────────────────────────────────────

    @Test void createWebhook_callsPost() {
        doReturn(HOOK_JSON).when(http).post(eq("/v1/webhooks"), any());
        api.createWebhook("https://example.com/hook", "booking.completed");
        verify(http).post(eq("/v1/webhooks"), any());
    }

    @Test void createWebhook_returnsParsedWebhook() {
        doReturn(HOOK_JSON).when(http).post(eq("/v1/webhooks"), any());
        XmlApiResponse<Webhook> resp = api.createWebhook("https://example.com/hook", "booking.completed");
        assertTrue(resp.isSuccess());
        assertEquals(1, resp.getData().getId());
    }

    // ── updateWebhook ──────────────────────────────────────────────────────────

    @Test void updateWebhook_callsPut_withId() {
        doReturn(HOOK_JSON).when(http).put(eq("/v1/webhooks/1"), any());
        api.updateWebhook(1, "https://new.example.com/hook", "booking.completed");
        verify(http).put(eq("/v1/webhooks/1"), any());
    }

    @Test void updateWebhook_differentId_usesCorrectPath() {
        doReturn(HOOK_JSON).when(http).put(eq("/v1/webhooks/99"), any());
        api.updateWebhook(99, "https://example.com/hook", "booking.cancelled");
        verify(http).put(eq("/v1/webhooks/99"), any());
    }

    // ── deleteWebhook ──────────────────────────────────────────────────────────

    @Test void deleteWebhook_callsDelete_withId() {
        doReturn(OK_JSON).when(http).delete("/v1/webhooks/1");
        api.deleteWebhook(1);
        verify(http).delete("/v1/webhooks/1");
    }

    @Test void deleteWebhook_returnsSuccess() {
        doReturn(OK_JSON).when(http).delete("/v1/webhooks/1");
        assertTrue(api.deleteWebhook(1).isSuccess());
    }

    // ── JAXB ───────────────────────────────────────────────────────────────────

    @Test void webhook_toXml_isValidXml() {
        Webhook w = new Webhook(); w.setId(1); w.setEvents(List.of("booking.completed")); w.setUrl("https://x.com");
        assertTrue(XmlConverter.isXml(XmlConverter.toXml(w)));
    }

    @Test void webhook_fromXml_roundTrip() {
        Webhook o = new Webhook(); o.setId(5); o.setEvents(List.of("booking.cancelled")); o.setUrl("https://t.com");
        Webhook r = XmlConverter.fromXml(XmlConverter.toXml(o), Webhook.class);
        assertEquals(5, r.getId());
        assertNotNull(r.getEvents());
        assertTrue(r.getEvents().contains("booking.cancelled"));
    }
}

File: kotlin-sdk/src/test/kotlin/com/cleanster/WebhooksTest.kt
package com.cleanster

import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class WebhooksTest {

    private lateinit var mock: MockHttpEngine
    private lateinit var client: CleansterClient

    @BeforeEach fun setUp() {
        mock   = MockHttpEngine()
        client = testClient(mock)
    }

    @Test fun `listWebhooks sends GET`() = runTest {
        mock.succeedList(); client.webhooks.listWebhooks()
        assertEquals("GET", mock.capturedMethod)
    }

    @Test fun `listWebhooks correct path`() = runTest {
        mock.succeedList(); client.webhooks.listWebhooks()
        assertTrue(mock.capturedUrl?.endsWith("/v1/webhooks") == true)
    }

    @Test fun `createWebhook sends POST`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://x.com", "event" to "booking.completed"))
        client.webhooks.createWebhook("https://x.com", "booking.completed")
        assertEquals("POST", mock.capturedMethod)
    }

    @Test fun `createWebhook correct path`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://x.com", "event" to "booking.completed"))
        client.webhooks.createWebhook("https://x.com", "booking.completed")
        assertTrue(mock.capturedUrl?.endsWith("/v1/webhooks") == true)
    }

    @Test fun `createWebhook encodes url`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://myapp.com/hook", "event" to "booking.created"))
        client.webhooks.createWebhook("https://myapp.com/hook", "booking.created")
        assertEquals("https://myapp.com/hook", mock.capturedBody?.get("url"))
    }

    @Test fun `createWebhook encodes event`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://x.com", "event" to "booking.cancelled"))
        client.webhooks.createWebhook("https://x.com", "booking.cancelled")
        assertEquals("booking.cancelled", mock.capturedBody?.get("event"))
    }

    @Test fun `createWebhook decodes id`() = runTest {
        mock.succeed(mapOf("id" to 99.0, "url" to "https://x.com", "event" to "booking.completed"))
        val resp = client.webhooks.createWebhook("https://x.com", "booking.completed")
        assertEquals(99, resp.data?.id)
    }

    @Test fun `updateWebhook sends PUT`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://new.com", "event" to "booking.started"))
        client.webhooks.updateWebhook(1, "https://new.com", "booking.started")
        assertEquals("PUT", mock.capturedMethod)
    }

    @Test fun `updateWebhook correct path`() = runTest {
        mock.succeed(mapOf("id" to 1.0, "url" to "https://new.com", "event" to "booking.started"))
        client.webhooks.updateWebhook(1, "https://new.com", "booking.started")
        assertTrue(mock.capturedUrl?.endsWith("/v1/webhooks/1") == true)
    }

    @Test fun `deleteWebhook sends DELETE`() = runTest {
        mock.succeedEmpty(); client.webhooks.deleteWebhook(1)
        assertEquals("DELETE", mock.capturedMethod)
    }

    @Test fun `deleteWebhook correct path`() = runTest {
        mock.succeedEmpty(); client.webhooks.deleteWebhook(1)
        assertTrue(mock.capturedUrl?.endsWith("/v1/webhooks/1") == true)
    }
}
